### PR TITLE
west: sign.py: add "REM" support to pass comments through cpp

### PR DESCRIPTION
**Pull request by marc-hb**
3 sign.py commits. Main commit:

west: sign.py: add "REM" support to pass comments through cpp

Generated outputs can be difficult to read, preserving comments helps a
lot and they often provide good `git grep` search keywords.

### CODE REVIEW

**Comment by marc-hb** [comment]
Successfully tested in commit https://github.com/marc-hb/sof/commits/toml-REM/

**Review by andyross — DISMISSED** [review]
No complaints from me, seems like a useful feature.  I will say though that this smells a little like a workaround for a design flaw: SOF, heh, sort of has a habit[1] of picking weak languages, trying to do complicated things with them, and then having to bolt on some kind of preprocessing to make up for the lack of programmability in the original.

So we have (here) rimage writing its config file in TOML, but then having to suck in non-trivial macrobatics to do what needs to be done, which requires debugging, which requires comments that survive the translation.

Maybe it might be time to think about generating the rimage config in a first class programming language (which you can simplify with a DSL layer as needed) rather than trying to build up toward a first class language with simpler tools?

[1] The true case study here was the attempt to bolt M4 around ALSA config files to generate topologies.  The tplg2 stuff walked that back, but now that too is being passed through cpp!

**Comment by marc-hb** [comment]
Thanks for the review!

>  but then having to suck in non-trivial macrobatics to do what needs to be done, which requires debugging, which requires comments that survive the translation.

"requires" sounds too strong; this PR is a "quality of life" enhancement.


> sort of has a habit[1] of picking weak languages,

Thanks for preaching to the choir (again) :-)
- https://github.com/zephyrproject-rtos/zephyr/pull/65411#issuecomment-1817304416
- https://github.com/thesofproject/rimage/pull/155#issuecomment-1499260734
- https://github.com/thesofproject/rimage/issues?q=crash
- + others if you follow these links

In theory, practice and theory are the same. In practice, people choose the languages they know and (think they) are the most productive with.

To be fair, this cpp-based solution seems very convenient to access `config.h` and similar. Zephyr is still 100% written in an ancient and weak language after all and it's not short of macrobatics either! "In Rome, do as the Romans do"

https://www.youtube.com/watch?v=w8GgP3h0M8M


Screenshot 2023-12-29 at 14 47 27


> The tplg2 stuff walked that back, but now that too is being passed through cpp!

I'm not familiar with topologies, do you have a link?

**Review by kv2019i — DISMISSED** [review]

File: scripts/west_commands/sign.py [hunk 1]
```diff
@@ -444,6 +444,12 @@ def preprocess_toml(self, config_dir, toml_basename, subdir):
         # unfortunately don't seem to make any difference here and they're
         # gcc-specific)
         preproc_cmd += ['-P']
+        # REM allows standalone comments to be preserved in generated/*.toml.
+        # A "double" `REM # my comment...` is not necessary but recommended for
+        # syntax highlighting.
+        # Note trailing comments are ignored by cpp and don't need any REM
+        # trick, a trailing `#` is enough.
+        preproc_cmd += ['-DREM=#']
```
Comment (lyakh) [inline scripts/west_commands/sign.py:452]: would it work to define `REM` an empty string and require "REM #" in toml files?

**Comment by nashif** [comment]
> Each platform has and always had its own build-platform/ build
directory. Keep all such build directories consistent with each other to
simplify installation, checksumming and post-processing. We don't need
any `build-mtl/**/mtl.toml` path to tell us the platform name _twice_,
once is enough.

> "Fixes" recent commit https://github.com/marc-hb/zephyr/commit/15336045aff85e2027748de89ab5a662492eff83 ("west: sign.py: generate platf.toml
from platf.toml.h with cc -E")


That commit message is confusing because it talks about how SOF builds things and not upstream Zephyr.

File: scripts/west_commands/sign.py (hunk 1 above)
Comment (marc-hb) [reply scripts/west_commands/sign.py:452]: OK, will do.

**Review by marc-hb — COMMENTED** [review]
> That commit message is confusing because it talks about how SOF builds things and not upstream Zephyr.

Thanks, I'll rephrase to be more neutral. Note the .toml pre-processing seems quite SOF-specific for now though.

**Comment by marc-hb** [comment]
v2; very small changes: 
- Commit message fix (@nashif)
- Replace `REM` words by nothing which makes `REM #` mandatory (@lyakh)

Approvals lost please review.

**Review by andyross — APPROVED** [review]
refresh +1

**Review by kv2019i — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**